services/venice-router/src/main/java/com/linkedin/venice/router/api/VenicePartitionFinder.java

```diff
@@ -106,15 +110,17 @@ private VenicePartitioner searchPartitioner(String storeName, int versionNum) {
     @Override
     public void handleStoreChanged(Store store) {
       String storeName = store.getName();
-      Set<Integer> upToDateVersionsSet =
-          store.getVersions().stream().map(Version::getNumber).collect(Collectors.toSet());
+      IntSet upToDateVersionsSet = store.getVersionNumbers();
 
       // remove out dated versions (if any) from the map
-      if (storeByVersionByPartitionerMap.containsKey(storeName)) {
-        Map<Integer, VenicePartitioner> versionByPartitionerMap = storeByVersionByPartitionerMap.get(storeName);
-        for (Integer candidateVersion: versionByPartitionerMap.keySet()) {
+      Map<Integer, VenicePartitioner> versionByPartitionerMap = storeByVersionByPartitionerMap.get(storeName);
+      if (versionByPartitionerMap != null) {
+        Iterator<Integer> candidateVersionIterator = versionByPartitionerMap.keySet().iterator();
+        Integer candidateVersion;
+        while (candidateVersionIterator.hasNext()) {
+          candidateVersion = candidateVersionIterator.next();
           if (!upToDateVersionsSet.contains(candidateVersion)) {
-            versionByPartitionerMap.remove(candidateVersion);
```

Comment:
Do we not still need to prune this?

Reply:
er wait, nevermind I think?  The candidateVersionIterator is based on this.  So I guess removing it from the iterator has the same effect?

Reply:
That's right, and from my understanding it is also safer to remove from the iterator...